### ISSUE TITLE
Proposal for a different modifier/stack structure

### ISSUE THREAD

**Issue by brnkhy** [mapbox/mapbox-unity-sdk#120]
We've been using ScriptableObjects for everything in mesh generation structure since the begining simply because they are awesome;
-serializable by default
-saves state even when changed during runtime
-easy to share, you can zip the whole folder and send your friend
-but most importantly, it fits well with the node based structure we had in mind since begining

it has some issues as well though;
-hard to navigate in&around modifier stack. Going back & switching between modifiers is hard without editor ui
-you end u with lots of asset files and it's hard to manage them as well. Sdk itself comes with lots of demo assets too so it's easy to find yourself in a position where you have multiple (but exactly same) modifiers as they don't have too many options to differentiate them either (see snap to terrian mod, polygon/line mods etc)
-we don't have too much customization in modifier at the moment so "sharing your modifier" is not a realistic scenario.
-makes it soo much harder to create multiple styles as you have to clone everything and rebind them etc.

We can do something different here; we can keep them as plain old (but serializable) C# classes instead of ScriptableObjects, then have ModifierStack (and rename that to Style) to manage them. So you'll end up with one asset file with one inspector to control all modifiers. It'll be so much easier to control the relationships between modifiers with that as well. 
And then this new thing, let's call it FeatureStyle (pretty much equivalent of the current modifier stack) would look like the post processing profile of Unity;
![unity_2017-06-09_17-32-01](https://user-images.githubusercontent.com/1465390/26980078-b3cde574-4d39-11e7-8e5b-1eb3d35c3b2a.png)
Imagine that with our modifiers.

I think this would solve all the problems above while keeping all the pros of scriptable objects, maybe except the node editor usability (it would be one big node in there, which beats the idea of node editor). but the best part will be the ease of navigation and creating new styles.

and actually, we probably keep both ModifierStack and this proposed FeatureStyle by wrapping that new  non-scriptable object modifier class with in a scriptable object as well. That might be pushing it too far though. 

@david-rhodes @isiyu @BergWerkGIS

**Comment by brnkhy** [comment]
OK so I've started working on this but I have some doubts about this now.
First of all, structure above will work with a fixed (and low) amount of modifiers but when you have 20 modifiers and want to add some custom stuff there, it'll get quite ugly. (Post processin asset has defined and low number of visual effets so it works for them)
That's why I tried to change it to an array based system where you would able to add and remove those modifiers but having issues with that as well. Unity serialization, uses the base class for modifiers (used for that array type) so serialized propertys can't reach inherited types (actual modifier code) properties/fields.
Post processing asset doesn't use array, probably for this reason and keeps all effects as individual fields of their own type so they don't run into this.

**Comment by brnkhy** [comment]
![unity_2017-06-10_01-59-51](https://user-images.githubusercontent.com/1465390/27043818-16951e22-4fa4-11e7-8cf8-749f27af6419.png)
this is what it looks like by the way

**Comment by mihakrajnc** [comment]
I like the concept, but as you said, custom modifiers can be a problem here and would get messy really fast (for example, I'm using 4 or 5 custom modifiers).

Also, this only solves a small part of the problem. As I understand it, you're basically rewriting the the modifier stack. There's still a lot of steps to go through before you get to the modifier stack (factories, layer visualizers etc...).

A good (but also complex) solution here (both functionally and visually) would be some sort of a node editor, similar to what Playmaker does. You would still have the same amount of `ScriptableObject`s in your folder structure (which the node editor might even organize and create for you?), but navigation your `MapVisualization` would be a lot easier and much more pleasing than it is now. It wouldn't even need to be editable, just having it display your MapVisualizer structure would be enough, and when you'd click on a node, that particular ScriptableObject would show up in the inspector.

~~No idea whatsoever how you'd put together a node editor in Unity (again i presume black magic and faries).~~ Actually, looking at [this example](https://forum.unity3d.com/threads/simple-node-editor.189230/), the concept of _making a node editor_ doesn't seem that complex. Adapting it for this is another matter, but still doable in my opinion.

Have you guys ever thought about anything like this? I might play around with it if I get the chance.

**Comment by brnkhy** [comment]
@mihakrajnc 
Post processing asset uses hard coded fields in code behind, which doesn't really work for our case (unknown amount of modifiers). I'm trying to change it to use array but turns out serialization part is tricky with that.
You probably already noticed but whole structure is actually built to be a node graph and used via a node editor since the begining. That's how I always visualize it in my head when I'm working and our direction will most likely be in that direction as well. This inspector stuff was supposed to be a short term solution but as you suggested, it might be easier to just switch to node editor at this point. If I remember correctly there're even some libraries for that on Github.

**Comment by mihakrajnc** [comment]
@brnkhy
It really is build as a node structure, kudos for that! Had no problem making a simple node viewer:

![image](https://user-images.githubusercontent.com/8857844/27058739-5634bc16-4fd3-11e7-9680-7cb7f9295291.png)

Just having this is already proving useful for navigating the top level elements.

I did find a library on github for this: https://github.com/Seneral/Node_Editor
Though I prefer the hands on "raw" example in this post: http://gram.gs/gramlog/creating-node-based-editor-unity/

I'll look at the node concept a bit more, it would be really useful to have an all out node editor implemented!

**Comment by brnkhy** [comment]
@mihakrajnc wow, that was quick! 👏 
I didn't know that second link you shared, looks like a good read, thanks.
I'll try to look into this as soon as I have some time. Please do share any progress you have though and don't hesitate to dm/mail whenever you need something or have questions

**Comment by brnkhy** [comment]
as a note, TreeView (https://docs.unity3d.com/Manual/TreeViewAPI.html) might be a quick solution to the navigation problem as well.

**Comment by david-rhodes** [comment]
We talked about this previously (before this repo was open-sourced), and here are some notes:

# TL;DR
Right now it is very cumbersome to edit the map visualization assets since you have to go back and forth between multiple objects in the project. The aim is to have the whole workflow in a single node editor window that gives you an easy overview of the whole hierarchy.		

# Functionality implemented in this ticket
- Visualize nodes
- Add nodes
- Remove nodes
- Edit nodes

# Ideas for implementation
- Have it mimic the Animation window as much as possible to provide Unity devs with a familar workflow and look.
- Use standard editors (Editor.CreateEditor) to draw the nodes for first iteration, use [HideInInspector] to not show the arrays that form the children of each node

# Open Questions
- Need a good way for each class in the hierarchy to inform the editor what its nodes are. 
- How to add a created child node to a node? Reflection? Or will switching to a common base class be enough? Tricky because some classes such as the MeshBuilder has more than one array child nodes are placed in.
- Do we have the nodes auto layout? Otherwise we will have to store their position in the graph view somewhere and it's probably best to do that inside the nodes themselves. This can either be done with a base class or an interface if having a common base class is not 
- When creating nodes do we still want them to exist as standalone asset files? It makes them easy to assign in the editor when using them for progress reporting for instance but it also means it's easy for the end user to delete them by accident.

https://cloud.githubusercontent.com/assets/1465390/21432797/224d7d34-c87e-11e6-8351-6f3d0ea77125.jpg

Yes, our original intent was always to enable developers to edit their map configurations inside a node editor, but we've been quite busy with other SDK improvements (performance and optimizations, mainly). We hope to tackle this soon.

Thanks for your assistance, @mihakrajnc!

**Comment by david-rhodes** [comment]
Node editor and asset selection browsers now in `develop`!
